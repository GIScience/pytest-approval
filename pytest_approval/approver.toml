approvers = [
    [
        "meld",
        "%received",
        "%approved"
    ],
    [
        "/usr/bin/flatpak",
        "run",
        "com.jetbrains.PyCharm-Community",
        "diff",
        "%received",
        "%approved"
    ],
    [
        "alacritty",
        "--command",
        "nvim",
        "-d",
        "%received",
        "%approved"
    ],
]
