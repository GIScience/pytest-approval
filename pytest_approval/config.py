import logging
import tomllib
from pathlib import Path


def _find_config(path: Path | None = None) -> Path | None:
    """Recursively walk up to root from current working dir to find pyproject.toml"""
    if path is None:
        path = Path.cwd()
    if str(path) == path.root:
        logging.debug("No pyproject.toml found.")
        return None
    config_path = path / "pyproject.toml"
    if config_path.exists():
        return config_path
    else:
        _find_config(path.parent)


def _read_config() -> dict:
    path = _find_config()
    if path is None:
        return {}
    with open(path, "rb") as file:
        config = tomllib.load(file)
    try:
        return config["tool"]["pytest-approval"]
    except KeyError:
        logging.debug("No configuration table found.")
        return {}
