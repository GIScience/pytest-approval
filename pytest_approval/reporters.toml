reporters = [
    [
        "diff",
        "--unified",
        "--color",
        "--suppress-common-lines",
        "--label",
        "received",
        "--label",
        "approved",
        "%received",
        "%approved",
    ],
]
