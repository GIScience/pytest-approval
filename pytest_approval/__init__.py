import logging

from pytest_approval.main import verify, verify_binary, verify_json

__all__ = ("verify", "verify_binary", "verify_json")

logging.basicConfig(
    level="DEBUG",
    format="%(asctime)s - %(levelname)s - %(filename)s - %(funcName)s - %(message)s",
)
