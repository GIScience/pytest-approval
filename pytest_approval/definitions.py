from contextlib import suppress
from pathlib import Path

from pytest_approval.config import _read_config

BASE_DIR = Path(__file__).parent.resolve()

REPORTERS_TEXT: list[list[str]] = [
    [
        "meld",
        "%received",
        "%approved",
    ],
    [
        "pycharm",
        "diff",
        "%received",
        "%approved",
    ],
    [
        "/usr/bin/flatpak",
        "run",
        "com.jetbrains.PyCharm-Community",
        "diff",
        "%received",
        "%approved",
    ],
    [
        "/usr/bin/code",
        "--new-window",
        "--wait",
        "--diff",
        "%received",
        "%approved",
    ],
    [
        "/usr/bin/open",
        # -W: Wait until the application is closed
        "-W",
        # -n: new instance
        "-n",
        # -a: application
        "-a",
        "/Applications/PyCharm Professional Edition.app/Contents/MacOS/pycharm",
        "--args",
        "diff",
        "%received",
        "%approved",
    ],
    [
        "/usr/bin/open",
        # -W: Wait until the application is closed
        "-W",
        # -n: New instance
        "-n",
        # -a: Application
        "-a",
        "/Applications/PyCharm CE.app/Contents/MacOS/pycharm",
        "--args",
        "diff",
        "%received",
        "%approved",
    ],
    [
        "/usr/bin/open",
        # -W: Wait until the application is closed
        "-W",
        # -n: New instance
        "-n",
        # -a: Application
        "-a",
        "/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code",
        "--args",
        "--new-window",
        "--wait",
        "--diff",
        "%received",
        "%approved",
    ],
    [
        "diff",
        "--unified",
        "--color",
        "--suppress-common-lines",
        "--label",
        "received",
        "--label",
        "approved",
        "%received",
        "%approved",
    ],
]


REPORTERS_BINARY: list[list[str]] = [
    [
        "pycharm",
        "diff",
        "%received",
        "%approved",
    ],
    [
        "/usr/bin/flatpak",
        "run",
        "com.jetbrains.PyCharm-Community",
        "diff",
        "%received",
        "%approved",
    ],
    [
        "diff",
        "--unified",
        "--color",
        "--suppress-common-lines",
        "--label",
        "received",
        "--label",
        "approved",
        "%received",
        "%approved",
    ],
]

BINARY_EXTENSIONS: list[str] = [
    # "7z",
    # "7zip",
    # "avif",
    # "bmp",
    # "bz2",
    # "bzip2",
    # "dds",
    # "dib",
    # "docx",
    # "emf",
    # "exif",
    # "gif",
    # "gz",
    # "gzip",
    # "heic",
    # "heif",
    # "ico",
    # "j2c",
    # "jfif",
    # "jp2",
    # "jpc",
    # "jpe",
    ".jpeg",
    ".jpg",
    # "jxr",
    # "nupkg",
    # "odp",
    # "ods",
    # "odt",
    # "pbm",
    # "pcx",
    # "pdf",
    # "pgm",
    ".png",
    # "ppm",
    # "pptx",
    # "rle",
    # "rtf",
    # "tar",
    # "tga",
    # "tif",
    # "tiff",
    # "wdp",
    # "webp",
    # "wmp",
    # "xlsx",
    # "xz",
    # "zip",
]

CONFIG = _read_config()

with suppress(KeyError):
    REPORTERS_TEXT = list(set(CONFIG["reporters"] + REPORTERS_TEXT))
