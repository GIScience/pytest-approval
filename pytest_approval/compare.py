import filecmp
from pathlib import Path
import logging

from pytest_approval.definitions import BINARY_EXTENSIONS


def compare_files(received: Path, approved: Path) -> bool:
    logging.debug(f"Compare {received} with {approved}.")
    if filecmp.cmp(received, approved, shallow=False):
        return True
    elif received.suffix not in BINARY_EXTENSIONS:
        return approved.read_text() == received.read_text()
    else:
        return False


def compare_image_contents_only(received: Path, approved: Path) -> bool:
    """Compare image contents without metadata."""
    try:
        import numpy
        from PIL import Image
    except ImportError as error:
        raise RuntimeError(
            'To use content_only, please install "pytest-approval[image]"'
            + '\n\n\tpip install "pytest-approval[image]"'
        ) from error
    received_image = Image.open(received)
    approved_image = Image.open(approved)
    return bool((numpy.array(received_image) == numpy.array(approved_image)).all())
