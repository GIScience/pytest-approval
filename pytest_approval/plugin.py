from pathlib import Path

from pytest_approval import main
from pytest_approval.definitions import CONFIG

auto_approve: bool = False


def pytest_addoption(parser):
    parser.addoption(
        "--auto-approve",
        action="store_true",
        help="Automatically approve every approval test",
    )


def pytest_configure(config):
    main.ROOT_DIR = config.rootpath
    main.AUTO_APPROVE = config.getoption("--auto-approve")
    approved_dir = CONFIG.get("approved-dir", None)
    if approved_dir is not None:
        main.APPROVED_DIR = approved_dir
        approved_dir_path = Path(main.ROOT_DIR) / Path(main.APPROVED_DIR)
        approved_dir_path.mkdir(parents=True, exist_ok=True)

def pytest_sessionfinish(session, exitstatus):
    main.cleaner(Path(main.ROOT_DIR) / Path(main.APPROVED_DIR))