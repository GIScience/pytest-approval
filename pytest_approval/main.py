import json
import logging
import os
import shutil
import subprocess
import zlib
from pathlib import Path
from typing import Any, Callable, Literal

from pytest_approval.compare import compare_files, compare_image_contents_only
from pytest_approval.definitions import (
    BASE_DIR,
    BINARY_EXTENSIONS,
    REPORTERS_BINARY,
    REPORTERS_TEXT,
)
from pytest_approval.utils import sort_dict

# will be instantiated during pytest configuration by plugin.py
ROOT_DIR: str = ""
APPROVED_DIR: str = ""
AUTO_APPROVE: bool = False

NAMES = []


def verify(
    data: Any,
    *,
    extension: str = ".txt",
) -> bool:
    return _verify(data, extension)


def verify_binary(
    data: Any,
    *,
    extension: Literal[".jpg", ".jpeg", ".png"],
) -> bool:
    return _verify(data, extension)


def verify_image(
    data: Any,
    *,
    extension: Literal[".jpg", ".jpeg", ".png"],
    content_only: bool = False,
) -> bool:
    """Verify image.

    Args:
        content_only: only compare content without metadata.
    """
    if content_only:
        return _verify(data, extension, compare=compare_image_contents_only)
    return _verify(data, extension)


def verify_json(
    data: str | dict,
    *,
    extension: Literal[".json"] = ".json",
    sort: bool = False,
) -> bool:
    if isinstance(data, str):
        data = json.loads(data)
    if sort:
        data = sort_dict(data)
    data = json.dumps(data, indent=True)
    return _verify(data, extension)


def _verify(data: Any, extension: str, compare: Callable = compare_files) -> bool:
    received, approved = _name(extension)
    _write(data, received, approved)
    if AUTO_APPROVE:
        shutil.copyfile(received, approved)
    if compare(received, approved):
        received.unlink()
        return True
    else:
        _report(received, approved)
        if compare(received, approved):
            received.unlink()
            return True
        else:
            return False


def _write(data, received, approved):
    """Write received to disk and create empty approved file if not exists."""
    if received.suffix in BINARY_EXTENSIONS:
        _write_binary(data, received, approved)
    else:
        _write_text(data, received, approved)


def _write_binary(data, received, approved):
    with open(received, "wb") as file:
        file.write(data)
    if not approved.exists():
        empty_file = Path(BASE_DIR / "empty_files" / "empty").with_suffix(
            approved.suffix
        )
        try:
            shutil.copy(empty_file, approved)
        except FileNotFoundError as e:
            raise ValueError(
                "Extension '{0}' not supported. ".format(approved.suffix)
                + "Extension for binary verification must be one of: {0}".format(
                    BINARY_EXTENSIONS
                )
            ) from e


def _write_text(data, received, approved):
    if len(data) == 0 or data[-1] != "\n":
        data = data + "\n"
    received.write_text(data)
    if not approved.exists():
        approved.touch()


def _name(extension=".txt") -> tuple[Path, Path]:
    # TODO: support base dir (then rewrite tests to use tmp_dir)
    # TODO: support postfix (write test with multiple calls to verify)
    # TODO: Try out with xdist
    node_id = os.environ["PYTEST_CURRENT_TEST"]
    if "[" in node_id and "]" in node_id:
        # TODO: Only use hash if params are loo long or special chars are presenet
        start = node_id.index("[") + 1
        end = node_id.index("]")
        params = node_id[start:end]
        hash = str(zlib.crc32(params.encode("utf-8")))
    else:
        params = ""
        hash = ""
    file_path = (
        node_id.replace(" (call)", "")
        .replace(" (setup)", "")
        .replace(" (teardown)", "")
        .replace("::", "--")
        .replace(params, hash)
    )
    count = _count(file_path)
    if APPROVED_DIR:
        file_path = Path(file_path)
        # find common parents between approved dir and file path, both
        # relative to pytest root, and remove them
        for i, part in enumerate(Path(APPROVED_DIR).parts):
            if part == file_path.parts[i]:
                continue
            else:
                break
        else:
            i = 0
        file_path = str(Path(*file_path.parts[i:]))
    received = file_path + count + ".received" + extension
    approved = file_path + count + ".approved" + extension
    return (
        Path(ROOT_DIR) / Path(APPROVED_DIR) / Path(received),
        Path(ROOT_DIR) / Path(APPROVED_DIR) / Path(approved),
    )


def _count(file_path: str) -> str:
    """Count generated names which are the same.

    This means `verify` has been called multiple times in one test function.
    """
    NAMES.append(file_path)
    count = NAMES.count(file_path)
    if count == 1:
        return ""
    else:
        return "." + str(count)


def _report(received: Path, approved: Path):
    if received.suffix in BINARY_EXTENSIONS:
        reporters = REPORTERS_BINARY
    else:
        reporters = REPORTERS_TEXT
    for command in reporters:
        command = [c.replace("%received", str(received)) for c in command]
        command = [c.replace("%approved", str(approved)) for c in command]
        try:
            completed_process = subprocess.run(  # noqa S603
                command,
                capture_output=True,
                check=False,
            )
            if completed_process.returncode == 127:  # command not found
                raise FileNotFoundError()  # noqa: TRY301
        except FileNotFoundError:
            logging.debug(f"Failed to run command `{' '.join(command)}` as approver.")
            continue
        if completed_process.returncode == 0:
            return
        elif completed_process.returncode == 1:
            msg = (
                "Received is different from approved.\n"
                + f"To approve run mv --force {received} {approved}\n"
            )
            raise AssertionError(msg + completed_process.stdout.decode("utf-8"))
        else:
            continue
    raise FileNotFoundError("No working approver could be found.")  # noqa: TRY003


def cleaner(path: Path):
    path = Path(path)
    if path.is_dir():
        for dirpath, _, filenames in os.walk(path):
            for filename in filenames:
                #filename = str(Path(filename).name)
                if ".approved." in filenames and filename not in NAMES:
                    file_path = os.path.join(dirpath, filename)
                    try:
                        os.remove(file_path)
                        print(f"Deleted: {file_path}")
                    except Exception as e:
                        print(f"Error deleting {file_path}: {e}")
