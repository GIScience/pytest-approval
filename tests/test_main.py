import logging
from pathlib import Path

import pytest

from pytest_approval import verify, verify_binary, verify_json
from pytest_approval.definitions import (
    BINARY_EXTENSIONS,
    REPORTERS_BINARY,
    REPORTERS_TEXT,
)
from pytest_approval.main import _name

FIXTURE_DIR = Path(__file__).parent / "fixtures"


@pytest.fixture
def approved():
    received, approved = _name()
    with open(approved, "w") as file:
        file.write("Hello World!\n")
    yield approved
    try:
        received.unlink()
        approved.unlink()
    except FileNotFoundError:
        pass


@pytest.fixture
def approved_different():
    received, approved = _name()
    with open(approved, "w") as file:
        file.write("hello world")
    yield approved


@pytest.mark.parametrize("string", ("Hello World!", "(id:(node/1, way/2))"))
def test_verify_string(string):
    assert verify(string)


# @pytest.mark.parametrize("reporter", REPORTERS_TEXT[:-1])
# @pytest.mark.parametrize(
#     "string",
#     ("", ":*?<>|\t\n\r\x0b\x0c"),
# )
# def test_verify_string(reporter, string, monkeypatch):
#     monkeypatch.setattr("pytest_approval.main.REPORTERS_TEXT", [reporter])
#     assert verify(string)


@pytest.mark.parametrize(
    "reporter",
    (REPORTERS_TEXT[0], REPORTERS_TEXT[2], REPORTERS_TEXT[3]),
)
def test_verify_string_all_reporter(reporter, monkeypatch):
    monkeypatch.setattr("pytest_approval.main.REPORTERS_TEXT", [reporter])
    assert verify("Hello World!")


@pytest.mark.parametrize(
    "json",
    (
        {"b": 100, "a": None},
        '{"b": 100, "a": null}',
    ),
)
def test_verify_json(json):
    assert verify_json(json)


def test_verify_json_sort():
    json = {"b": 100, "a": {"d": 10, "c": 10}}
    assert verify_json(json, sort=True)


# TODO read empty files for extension and verify it:
@pytest.mark.parametrize("extension", BINARY_EXTENSIONS)
def test_verify_binary(extension, monkeypatch):
    monkeypatch.setattr("pytest_approval.main.REPORTERS_BINARY", [REPORTERS_BINARY[1]])
    with open(FIXTURE_DIR / f"binary{extension}", "rb") as file:
        data = file.read()

    assert verify_binary(data, extension=extension)


def test_verify_gnu_diff_tools_approver(monkeypatch):
    monkeypatch.setattr("pytest_approval.main.REPORTERS_TEXT", [REPORTERS_TEXT[-1]])
    with pytest.raises(AssertionError) as error:
        assert verify("Hello World!")
    expected = r"""Received is different from approved.
To approve run mv --force /home/matthias/projects/pytest-approval/tests/test_main.py--test_verify_gnu_diff_tools_approver.received.txt /home/matthias/projects/pytest-approval/tests/test_main.py--test_verify_gnu_diff_tools_approver.approved.txt
--- received
+++ approved
@@ -1 +0,0 @@
-Hello World!
"""  # noqa
    assert expected == str(error.value)


def test_verify_approved_equal(approved, fake_process):
    fake_process.register_subprocess(["meld", fake_process.any()])
    assert verify("Hello World!") is True
    assert fake_process.call_count(["meld", fake_process.any()]) == 0


def test_verify_approved_different(approved_different, fake_process):
    fake_process.register_subprocess(["meld", fake_process.any()])
    assert verify("Hello World!") is False
    assert fake_process.call_count(["meld", fake_process.any()]) == 1


def test_verify_approved_none(fake_process):
    fake_process.register_subprocess(["meld", fake_process.any()])
    assert verify("Hello World!") is False
    assert fake_process.call_count(["meld", fake_process.any()]) == 1


def test_verify_different_returncode_127(fake_process, caplog):
    caplog.set_level(logging.DEBUG)
    fake_process.register_subprocess(
        ["meld", fake_process.any()],
        returncode=127,
    )
    fake_process.register_subprocess(
        ["pycharm", fake_process.any()],
        returncode=0,
    )
    assert verify("Hello World!") is False
    assert "Failed to run command" in caplog.text
    assert fake_process.call_count(["meld", fake_process.any()]) == 1
    assert fake_process.call_count(["pycharm", fake_process.any()]) == 1
