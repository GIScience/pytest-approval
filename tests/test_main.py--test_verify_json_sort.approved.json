{"a": {"c": 10, "d": 10}, "b": 100}
